File: leetcode/Stack/leetcode-32/32.java
class Solution {
    public int longestValidParentheses(String s) {
        if (s == null || s.length() < 2) {
            return 0;
        }

        Stack<Integer> stack = new Stack<>();

        int start = 0;
        int res = 0;

        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                stack.push(i);
            } else {
                if (stack.isEmpty()) {
                    start = i + 1;
                } else {
                    int tmp = stack.pop();
                    res = stack.isEmpty() ? Math.max(res, i - start + 1) : Math.max(res, i - stack.peek());
                }
            }
        }
        return res;
    }
}

File: leetcode/Stack/leetcode-32/32.go
package stack

import "container/list"

func longestValidParentheses(s string) int {
	if len(s) < 2 {
		return 0
	}

	start, res := 0, 0
	stack := list.New()

	for i := 0; i < len(s); i++ {
		if s[i] == '(' {
			stack.PushBack(i)
		} else {
			if stack.Len() == 0 {
				start = i + 1
			} else {
				stack.Remove(stack.Back())
				if stack.Len() == 0 {
					res = max(res, i-start+1)
				} else {
					res = max(res, i-stack.Back().Value.(int))
				}
			}
		}
	}
	return res
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
